### DISCUSSION TITLE
Does Autogluon have automated feature selection

### DISCUSSION THREAD

**Discussion by apavlo89** [autogluon/autogluon#3831] [Q&A] [answered]
Does it?

**Comment by mglowacki100** [comment, marked as answer]
Yes, but it is experimental feature disabled by default. To activate it in method `.fit(feature_prune_kwargs={'force_prune': True})`

**Reply by apavlo89** [reply]
Wow, i really expected the answer to be negative! What feature selection algorithm is it using and does it perform well under high dimensionality?

**Reply by apavlo89** [reply]
Wait, I just run the thing no feature selection was applied. No output describing feature selection either.

4912 features in original data used to generate 4912 features in processed data.
	Train Data (Processed) Memory Usage: 65.36 MB (0.1% of available memory)
Data preprocessing and feature engineering runtime = 37.14s ...
AutoGluon will gauge predictive performance using evaluation metric: 'balanced_accuracy'
	To change this, specify the eval_metric parameter of Predictor()
Large model count detected (112 configs) ... Only displaying the first 3 models of each family. To see all, set `verbosity=3`.

**Reply by shchur** [reply]
Feature selection is done after all models are fit using permutation testing. You can check out the documentation for `feature_prune_kwargs ` at https://auto.gluon.ai/stable/api/autogluon.tabular.TabularPredictor.fit.html for more details

**Reply by apavlo89** [reply]
@shchur @mglowacki100  Great so it is done after all models are fit. I am worried about using permutation testing with almost 5k feature and I'd rather not use permutation importance if possible.  Is there a way to use the model parameters that autogluon was trained on wrapped on a feature selection algorithm of my choice?

I've got this awesome feature selection algorithm (https://github.com/Ahadzadeh2022/SFE/tree/main | https://ieeexplore.ieee.org/document/10024844) that has been performing really well in high dimensional databases with various classifiers, however I am unsure how to load the autogluon model with its parameters to be used

**Reply by jmren168** [reply]
@apavlo89 

You could try 
`hyperparameters.update({'GBM':{}})`

`feature_prune_kwargs = {
            'n_train_subsample': 50000,
            'n_fi_subsample': 10000,
            'prune_ratio': 0.05,
            'stopping_round': 10,
            'min_improvement': 1e-6,
            'force_prune': True,
        }`

train Autogluon model
`predictor = TabularPredictor(label=label, path=save_path).fit(
                                            train_data,
                                            hyperparameters=hyperparameters, 
                                            hyperparameter_tune_kwargs=hyperparameter_tune_kwargs,
                                            feature_prune_kwargs=feature_prune_kwargs,
                                            verbosity=4)`

However, GBM will be one base models when performing ensemble.

@shchur @mglowacki100 
Is it possible to remove GBM when performing ensemble? In other words, GBM is only used for pruning features. Many thanks.

**Reply by shchur** [reply]
I'm not that familiar with how feature pruning works in `autogluon.tabular`, @Innixma might have more context on this.

Some ideas: 
- `excluded_model_types` argument to `TabularPredictor.fit()` allows to exclude some models from training, but I'm not sure if this won't prevent GBM from being used for feature selection. 
- You could try splitting the whole process into two parts: first creating one predictor that fits only GBM to perform feature pruning, then training a new predictor on the subset of selected features.

**Reply by Innixma** [reply]
Note that I didn't benchmark feature pruning for AG 1.0. It is only used when enabled if all models are fit and your time limit is above half its original value, as it will then re-train all of the models again with the pruned features. Because `best_quality` now uses Zeroshot-HPO portfolio of 113 models, that means by default it will need to train all 113 models in under half the time limit to perform feature pruning if you enable it.

At some point I intend to refactor this logic, as it made a lot more sense when we had 13 models, but is less practical when we have 113 models. Alternatives could be to train models in a layer until half of time is used, then feature prune, then train as many feature pruned models as possible. Would require some code edits in AbstractTrainer class, and has a strong chance of worse results in low time limit scenarios, so I'd be careful of enabling it by default.

**Reply by Innixma** [reply]
Regarding dropping in a custom feature importance logic, you'd need to edit the code currently. This was an intern project from 2021 and we haven't gone back and made it fully extensible / modular yet. This is a big reason it isn't enabled by default, because doing iterative feature elimination via permutation shuffling on 10K+ features can take significant amounts of time, and I wanted to avoid users having to wait unreasonably long in those cases.

**Reply by jmren168** [reply]
@shchur @Innixma 
Thank you for the reply. It's a good idea to split the whole process into two parts.

**Reply by jmren168** [reply]
@shchur 

Any ideas of how to get the pruned features?

I tried to use "predictor._learner.feature_generator.transform(train_data)", but still got the original feature set.
Or how to load a LightGBM\\T1_Prune\\model.pkl to get the used features?

Mank thanks.

**Reply by jmren168** [reply]
It seems like predictor does not save the pruned features. fit_summary is showed as bellow.

```
*** Summary of fit() ***
Estimated performance of each model:
                 model  score_val eval_metric  pred_time_val   fit_time  pred_time_val_marginal  fit_time_marginal  stack_level  can_infer  fit_order
0    LightGBM\T5_Prune   0.998457    accuracy       0.011999   5.186092                0.011999           5.186092            1       True          5
1    LightGBM\T1_Prune   0.998457    accuracy       0.013999   8.284000                0.013999           8.284000            1       True          1
2  WeightedEnsemble_L2   0.998457    accuracy       0.016000   8.967139                0.002001           0.683138            2       True          6
3    LightGBM\T3_Prune   0.996914    accuracy       0.017001   9.069832                0.017001           9.069832            1       True          3
4    LightGBM\T4_Prune   0.996914    accuracy       0.084004  60.405211                0.084004          60.405211            1       True          4
5    LightGBM\T2_Prune   0.992284    accuracy       0.012002   9.139832                0.012002           9.139832            1       True          2
Number of models trained: 6
Types of models trained:
{'WeightedEnsembleModel', 'LGBModel'}
Bagging used: False 
Multi-layer stack-ensembling used: False 
Feature Metadata (Processed):
(raw dtype, special dtypes):
('float', []) : 220 | ['a','b','c', ...]
*** End of fit() summary ***
```

I'm sure the feature selection is successfully pruning 220 features to 141 features, see log as below

	Successfully ended prune loop after 15 feature pruning rounds (229.08s).
	Feature count before/after feature pruning: 220 -> 141.

**Reply by jia-zhuang** [reply]
Just load the model.pkl, you can get the features:
```python
import pickle
with open('AutogluonModels/ag-20240510_135549/models/LightGBM_Prune/model.pkl', 'rb') as f:
    model = pickle.load(f)

print(model.features)
```

**Reply by jmren168** [reply]
@jia-zhuang 

Thanks. It works.
